<a href="https://colab.research.google.com/github/stefkong1982/netology.ru/blob/Master/Statistika_v_Python/Sluchajnye_sobytiya_Sluchajnye_velichiny/DZ_Kondrstev_STPY_48_Sluchajnye_velichiny.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Домашнее задание к лекции «Случайные события. Случайные величины»


Задание
Вопросы по заданию
Преподаватель: Олег Булыгин
Домашнее задание к лекции «Случайные события. Случайные величины»

Смоделировать игру против лотерейного автомата типа “777”. Игрок платит 1 руб., после чего выпадает случайное целое число, равномерно распределенное от 0 до 999. При некоторых значениях числа игрок получает выигрыш.

777: 200 руб.

999: 100 руб.

555: 50 руб.

333: 15 руб.

111: 10 руб.

*77: 5 руб.

**7: 3 руб.

*00: 2 руб.

**0: 1 руб.

Вопросы:

Выгодна ли игра игроку?
Сколько в среднем приобретает или теряет игрок за одну игру?

**Дополнительное задание повышенной сложности.**

Теоретически рассчитать средний выигрыш (проигрыш) и сравнить с результатами моделирования

`*` любая цифра

Вероятности рассчитываются исходя из общего числа вариантов, равного 1000 (числа от 000 до 999).

1. Выпадение числа 777, 999, 555, 333, 111 каждое из этих чисел — это 1 уникальный исход из 1000 возможных. Следовательно, вероятность выпадения каждого из этих чисел равна 1/1000, или 0,1%.

2. Выпадение числа *77 — здесь звездочка обозначает любую цифру. Мы имеем 10 возможных исходов (цифры от 0 до 9), в каждом из которых последние две цифры — 77. То есть вероятность такого исхода — 10/1000, или 1%.

3. Выпадение числа **7 — две звездочки обозначают любые две цифры.  Здесь мы имеем 100 возможных исходов (числа от 00 до 99), в каждом из которых последняя цифра — 7. То есть вероятность такого исхода — 100/1000, или 10%.

4. Выпадение числа *00 аналогично случаю с числом *77, количество возможных исходов — 10, и вероятность выпадения составит 1%.

5. Выпадение числа **0 здесь имеем 100 возможных исходов и вероятность выпадения также составит 10%.

In [ ]:
total_outcomes = 1000    # Всего возможных исходов
cost_of_game = 1         # Стоимость игры

# список чисел и их выплат
numbers_and_payouts = {'777': 200, '999': 100, '555': 50, '333': 15, '111': 10, '*77': 5, '**7': 3, '*00': 2, '**0': 1}

# Вероятности для каждой выплаты
probabilities = {'777': 1/total_outcomes, '999': 1/total_outcomes,
                 '555': 1/total_outcomes, '333': 1/total_outcomes,
                 '111': 1/total_outcomes, '*77': 10/total_outcomes,
                 '**7': 100/total_outcomes, '*00': 10/total_outcomes,
                 '**0': 100/total_outcomes}

# Считаем средний выигрыш или проигрыш
expected_value = 0
for number, payout in numbers_and_payouts.items():
    expected_value += payout * probabilities[number]
expected_loss = cost_of_game - expected_value

print(f"Средний выигрыш: {expected_value} руб.")
print(f"Средний проигрыш: {expected_loss} руб.")


Средний выигрыш: 0.8450000000000001 руб.
Средний проигрыш: 0.15499999999999992 руб.


В этом коде для каждой выплаты мы вычисляем ожидаемую величину выигрыша или проигрыша, умножая выплату на соответствующую вероятность, и суммируем результаты. Потом вычитаем стоимость игры из общего среднего выигрыша, чтобы узнать средний проигрыш.

1. Вероятность выиграть 200 руб. (выпадет 777) — 0.001 (или 0.1%), значит с каждого рубля игрок в среднем выиграет 0.001*200 = 0.2 руб.

2. Вероятность выиграть 100 руб. (выпадет 999) — 0.001, значит с каждого рубля игрок в среднем выиграет 0.001*100 = 0.1 руб.

3. И так далее для остальных чисел: Выпадение числа 555 — 0.05 руб., 333 — 0.015 руб., 111 — 0.01 руб., *77 — 0.05 руб., **7 — 0.3 руб., *00 — 0.02 руб., **0 — 0.1 руб.

Суммируя все эти средние выигрыши, получим теоретическую среднюю сумму, которую игрок выигрывает с каждого потраченного рубля:

0.2 + 0.1 + 0.05 + 0.015 + 0.01 + 0.05 + 0.3 + 0.02 + 0.1 = 0.845 руб.

То есть в долгосрочной перспективе, теоретически, за каждый потраченный рубль игрок получает 0.845 руб.

Таким образом, исходя из нашего предположения, что игра стоит 1 рубль, в среднем игрок проиграет ~0.155 рубля с каждой игры.

### Рассчитаем вероятность из моделирования

In [ ]:
import random

total_games = 1000000    # общее количество игр
cost_of_game = 1         # стоимость игры
total_cost = total_games * cost_of_game  # общая стоимость всех игр

# проверка номера и определение выплаты
def check_number(num):
    if num == 777: return 200
    elif num == 999: return 100
    elif num == 555: return 50
    elif num == 333: return 15
    elif num == 111: return 10
    elif num % 100 == 77: return 5
    elif num % 10 == 7: return 3
    elif num % 100 == 0: return 2
    elif num % 10 == 0: return 1
    else: return 0

total_win = sum(check_number(random.randint(0, 999)) for _ in range(total_games))  # общий выигрыш

avg_win = total_win / total_games  # средний выигрыш на игру
avg_loss = cost_of_game - avg_win  # средний проигрыш на игру

print(f"Средний выигрыш: {avg_win} руб.")
print(f"Средний проигрыш: {avg_loss} руб.")


Средний выигрыш: 0.791326 руб.
Средний проигрыш: 0.20867400000000003 руб.


В этом коде мы запускаем `total_games` игр (`total_games = 1000000`). Для каждой игры мы случайно генерируем число от 0 до 999 и проверяем его на соответствие условиям выплат. Затем мы суммируем все выплаты, чтобы получить общий выигрыш.

Средний выигрыш на игру вычисляется как общий выигрыш, поделенный на общее количество игр. Средний проигрыш на игру вычисляется как стоимость игры минус средний выигрыш на игру.

Разница в результатах от теоретического расчета и смоделированного эксперимента объясняется случайными колебаниями в эксперименте и ограниченностью числа попыток.

В теоретическом расчете мы использовали точные вероятности для каждого исхода (которые были определены заранее) и посчитали ожидаемый выигрыш по формулам вероятности. Этот расчет предполагает бесконечное количество попыток.

С другой стороны, при моделировании мы провели ограниченное количество попыток (в примере 1 миллион). Несмотря на то, что это исчисляется миллионами, оно все же является конечным числом и, следовательно, подвержено случайным отклонениям. Результат каждого отдельного запуска может немного отклоняться от ожидаемого значения из-за случайности, вмешивающейся в процесс.

В долгосрочной перспективе, при бесконечном количестве попыток, средний выигрыш, полученный с помощью моделирования, будет все ближе и ближе к теоретически рассчитанному выигрышу.